rustdoc: Align search results horizontally for easy scanning

camelid opened the pull request:
The recent PR #110688 added info about an item's kind before its name in
search results. However, because the kind and name are inline with no
alignment, it's now hard to visually scan downward through the search
results, looking at item names. This PR fixes that by horizontally
aligning search results such that there are now two columns of
information.

r? @GuillaumeGomez

camelid committed:
Align search results horizontally for easy scanning

The recent PR #110688 added info about an item's kind before its name in
search results. However, because the kind and name are inline with no
alignment, it's now hard to visually scan downward through the search
results, looking at item names. This PR fixes that by horizontally
aligning search results such that there are now two columns of
information.

compiler-errors commented:
Do you have a rendering of what the new results look like?

Also, quite in favor of this change 👍

camelid commented:
Note: I'm having an issue getting Rust to build on my computer right now, so the "after" screenshot was generated by applying the changes in Firefox dev tools.

### Before

![image](https://github.com/rust-lang/rust/assets/37223377/283adb89-d976-4bf2-9809-72b543b08fbb)


### After

![image](https://github.com/rust-lang/rust/assets/37223377/7e8944a3-275a-428b-bfdd-e8f097e821dc)

camelid commented on src/librustdoc/html/static/css/rustdoc.css:
```diff
@@ -891,8 +891,10 @@ so that we can apply CSS-filters to change the arrow color in themes */
+	display: inline-block;
 	color: var(--search-results-grey-color);
 	font-size: 0.875rem;
+	width: 6rem;
```
I picked this number arbitrarily. Two issues:

1. It would be nice if it weren't based on eyeballing the layout.

2. This isn't actually wide enough for "associated constant". If I increase this to 9em, it fits, but then all the other rows have excessive spacing. I see two solutions: one is to shorten the typename to just "constant", the other is to make the spacing somehow adaptive to what the longest typename is. But I'm not sure how to implement that.

compiler-errors replied:
What happens when it overflows? Does it wrap and just become a taller row? I think that honestly wouldn't be terrible if that's what happens.

camelid replied:
It's not terrible, but still looks bad IMO:

![image](https://github.com/rust-lang/rust/assets/37223377/de69e7a9-934d-4dc5-821a-1f49ee05a316)

compiler-errors replied:
:/ true -- i wonder if vertically center-aligning those rows would help.

camelid replied:
I think we really need to just keep rows a constant height. Variable height rows almost always look a bit off.

albertlarsan68 replied:
Something that could at least make them look better is to make sure both the path/name and short description are aligned the same, to the top or the middle as Michael suggested.

camelid replied:
I was able to avoid the wrapping by adding a tiny bit extra space.

GuillaumeGomez commented:
The longest type name is `associated constant`. So `calc(0.875rem * 20)` should do the trick. But it's quite wide, hence why we finally decided to go without it. UI is tricky. ^^'

camelid commented:
@GuillaumeGomez could we shorten "associated constant" to just "constant"? I'm not sure if it really matters to users whether a constant is associated or free, and even aside from this change, it's quite a long word that distracts from the actual name of the item.

compiler-errors commented:
> could we shorten "associated constant" to just "constant"?

I'd honestly rather shorten it to `assoc const` or something. At least on the compiler side, we *try* to be consistent with calling things "associated".

camelid commented:
Abbreviation would be fine with me 👍

GuillaumeGomez commented:
But `associated const` is already super long (especially if no type kind this long is displayed), and also it would become the second longest after `existential type`. But we can already try to see how it looks with `associated const` already?

camelid committed:
Abbreviate long typenames so they don't get wrapped in results

camelid commented:
I abbreviated associated type too (not visible here):

![image](https://github.com/rust-lang/rust/assets/37223377/4231a8ba-11c0-4b5f-a56d-45dbf652d7ef)

GuillaumeGomez commented:
It looks nice like that. But remains the problem of `existential type`.

camelid commented:
Existential types have not been implemented yet I thought? Or are at least I thought an experimental feature.

camelid commented:
"Existential type" is still shorter than "associated constant". We could shorten "existential type" to just "existential"?

What is an existential type in this context actually? Is it `type Foo = impl Trait`?

compiler-errors commented:
@camelid: type alias impl trait ("existential type" in rustdoc i guess) is implemented, and we probably should account for them too

camelid commented:
We'd actually only need 7 rem (rather the current 6) to support "existential type" as it is.

camelid committed:
Make typenames a bit wider to support "existential type"

GuillaumeGomez commented:
Why `7rem`? "existential type" is 17 characters. So `17 * 0.875 = 14.875rem`, not 7. Did I miss something?

camelid commented:
In non-monospace fonts, the letters are variable width, so you can't just multiply the width. In addition, I believe the "em" in "rem" is referring to a value that is (approximately, sigh...) the width of an uppercase "M".

camelid commented:
New screenshot. Let me know if there are any other types of items that you'd like to see rendered.

![image](https://github.com/rust-lang/rust/assets/37223377/84448248-fca3-42fa-969f-ec3cc75d5a4f)

GuillaumeGomez commented:
Can you just check what it gives with "existential type" please? (Just update the text, no need for a real one :smile_cat: )

Also please show screenshots of a mobile view (just reduce the width).

camelid commented:
By the way, I added a demo to my "rem" explanation above in case you're interested :)

"existential type":

![image](https://github.com/rust-lang/rust/assets/37223377/2324d13d-7295-4a6d-ab59-fe0eae495cb1)

mobile view:

![image](https://github.com/rust-lang/rust/assets/37223377/79b897dd-ffa0-4659-aaf7-327ff66b29d1)

camelid commented:
Some options:

### 6.25 rem

IMO the "existential type" is too close to the name.

![image](https://github.com/rust-lang/rust/assets/37223377/d491fc4c-46dc-492b-9b08-88c41206ce1e)

### 6.5 rem

![image](https://github.com/rust-lang/rust/assets/37223377/f987702c-6415-4575-9315-065c4f936774)

### 6.75 rem

![image](https://github.com/rust-lang/rust/assets/37223377/4172b78d-437c-4bb0-b6e0-2db4fd70223d)

### 7 rem

![image](https://github.com/rust-lang/rust/assets/37223377/bb166ccd-6cb8-4522-82ef-bf112fc050f8)

camelid commented:
I think I like 6.5 rem the best. Here's what it looks like without "existential type":

![image](https://github.com/rust-lang/rust/assets/37223377/eba51d9c-ff29-4a98-9972-964771a02e76)

GuillaumeGomez commented:
I vote for 6.25. :)

camelid commented:
I agree with you about 6.25rem being the best. My only concern is it looks cramped with "existential type". So which do you think is best overall given that tradeoff? :)

GuillaumeGomez commented:
I still prefer 6.25. :p

camelid committed:
Reduce typename width to 6.25rem

This makes "existential type" look slightly cramped (though still
readable), but it makes all other typenames look better. Existential
types are currently very rare, and we can always tweak this later if
necessary.

camelid commented:
Ok, here is the final version (6.25rem):

![image](https://github.com/rust-lang/rust/assets/37223377/bf79b60b-11e3-4872-bba9-9ea660a4a5e9)

---

And with an artificial "existential type":

![image](https://github.com/rust-lang/rust/assets/37223377/f88a9343-30d7-4ca9-a7ca-42108cd82f66)

GuillaumeGomez commented:
Looks good to me. Well done! :+1:

GuillaumeGomez commented:
Does it look good to you @compiler-errors ? Also cc @notriddle

compiler-errors commented:
👍

camelid commented:
@bors r=GuillaumeGomez

bors commented:
:pushpin: Commit 5f433f33ed8d4a8ede25c36d2bb0543e8b101920 has been approved by `GuillaumeGomez`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=GuillaumeGomez 5f433f33ed8d4a8ede25c36d2bb0543e8b101920 -->
<!-- homu: {"type":"Approved","sha":"5f433f33ed8d4a8ede25c36d2bb0543e8b101920","approver":"GuillaumeGomez","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

camelid commented:
@bors rollup

GuillaumeGomez commented:
Failed in https://github.com/rust-lang/rust/pull/113004#issuecomment-1605676358. You forgot to update a GUI test. ;)

@bors r-

camelid committed:
rustdoc: Update GUI test

GuillaumeGomez commented:
Thanks!

@bors r=notriddle,GuillaumeGomez rollup

bors commented:
:pushpin: Commit 9b97ae1d8c6e99a4832c84dec4d4b6faf5f583f6 has been approved by `notriddle,GuillaumeGomez`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=notriddle,GuillaumeGomez 9b97ae1d8c6e99a4832c84dec4d4b6faf5f583f6 -->
<!-- homu: {"type":"Approved","sha":"9b97ae1d8c6e99a4832c84dec4d4b6faf5f583f6","approver":"notriddle,GuillaumeGomez","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #113006

bors commented:
:pushpin: Commit 9b97ae1d8c6e99a4832c84dec4d4b6faf5f583f6 has been approved by `notriddle,GuillaumeGomez`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=notriddle,GuillaumeGomez 9b97ae1d8c6e99a4832c84dec4d4b6faf5f583f6 -->
<!-- homu: {"type":"Approved","sha":"9b97ae1d8c6e99a4832c84dec4d4b6faf5f583f6","approver":"notriddle,GuillaumeGomez","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
